### PR TITLE
[core][dashboard] Task backend GC policy - worker update [1/3]

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This is the series of PRs that improve GC policy for task backend. The overall goal of the stack is to make data loss at a task attempt granularity: if a task attempt incurred some data loss (due to number of task events enforced at the worker/ GCS), all the status change w.r.t that task attempt will be dropped, so there will be no partial task attempt. 
Right now, individual events (e.g. task started running) could be lost for a task attempt, which isn't great for observability. 

This PR adds task attempt level data loss tracking on the worker side, by tracking:
1. per job profile events dropped for timeline.
2. task attempts dropped. 

Worker will send the data loss info ^ the GCS 

In the subsequent PRs:
- GCS side will be updated
- Dashboard front-end will be updated to reflect the per job profile events loss. 

## Checks

Release tests here on microbenchmark + stress_test_many_tasks no regression: https://buildkite.com/ray-project/release-tests-pr/builds/36724#_

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/core_worker/task_event_buffer.h [hunk 1]
```diff
@@ -173,13 +177,23 @@ enum TaskEventBufferCounter {
...
 ///
 ///   2. More than `RAY_task_events_max_buffer_size` tasks have been stored
-///   in the buffer, any new task events will be dropped. In this case, the number of
-///   dropped task events will also be included in the next flush to surface this.
+///   in the buffer, oldest events in the buffer will be dropped. In this case, the task
+///   attempts info will also be included in subsequent flush to GCS.
+///
+/// For profiling events:
+///   - If the number of profiling events for a task attempt exceeds the limit specified
+///   by `RAY_task_max_profile_events_per_task`, any new profiling events will be dropped.
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:190]: Is it the correct name? I couldn't find it from the config

File: src/ray/core_worker/task_event_buffer.h [hunk 2]
```diff
@@ -325,6 +327,14 @@ class TaskEventBufferImpl : public TaskEventBuffer {
   /// process them quick enough.
   std::atomic<bool> grpc_in_progress_ = false;
 
+  /// Profile events dropped per job, to be reported to GCS. Reported data loss
+  /// will be removed from this.
+  absl::flat_hash_map<JobID, uint32_t> profile_events_dropped_ GUARDED_BY(mutex_);
```
Comment (rkooo567) [inline src/ray/core_worker/task_event_buffer.h:332]: We don't need to categorize this per job id. There can be only 1 job per worker.

File: src/ray/gcs/gcs_server/gcs_task_manager.h [hunk 3]
```diff
@@ -277,6 +277,8 @@ class GcsTaskManager : public rpc::TaskInfoHandler {
   };
 
  private:
+  void RecordDataLossFromWorker(const rpc::TaskEventData &data);
```
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_task_manager.h:280]: Can you add a docstring?

**Review by rkooo567 — APPROVED** [review]
LGTM. Some nit comments.

Btw, also a Q: Have you seen practically the cases where the task events are lost from the worker side? It seems really rare.

File: src/ray/core_worker/task_event_buffer.h (hunk 2 above)
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.h:332]: ah - right. Good catch.

**Comment by rickyyx** [comment]
> LGTM. Some nit comments.
> 
> Btw, also a Q: Have you seen practically the cases where the task events are lost from the worker side? It seems really rare.

it should be rather rare, I believe only when: 
- There are large number of workers so GCS might be slow processing task events from all workers so workers stop flushing. 
- AND a large number of tasks being submitted to the worker. 

I can look into a few release tests. The last time I think it was more in the scale of 100K+ backlog.

File: src/ray/core_worker/task_event_buffer.h (hunk 1 above)
Comment (rickyyx) [reply src/ray/core_worker/task_event_buffer.h:190]: task_max_profile_events_per_task

**Pull request merged**